File: cpp/build-company/Equipment.cpp
#include "Equipment.hpp"
#include <iostream>

// Реализация конструктора
Equipment::Equipment(const std::string& name, int count)
    : name(name), count(count) {
}

// Реализация деструктора
Equipment::~Equipment() {
    std::cout << "Вызван деструктор Equipment для: " << name << std::endl;
}

// Реализация геттеров
std::string Equipment::getName() const {
    return name;
}

int Equipment::getCount() const {
    return count;
}

// Реализация сеттеров
void Equipment::setName(const std::string& name) {
    this->name = name;
}

void Equipment::setCount(int count) {
    this->count = count;
}

// Добавление количества
void Equipment::addCount(int amount) {
    if (amount > 0) {
        count += amount;
        std::cout << "Добавлено " << amount << " ед. " << name
                  << ". Новое количество: " << count << std::endl;
    }
}

// Удаление количества
void Equipment::removeCount(int amount) {
    if (amount > 0 && amount <= count) {
        count -= amount;
        std::cout << "Удалено " << amount << " ед. " << name
                  << ". Осталось: " << count << std::endl;
    } else if (amount > count) {
        std::cout << "Ошибка: Невозможно удалить " << amount << ". Доступно только " << count << std::endl;
    }
}

File: csharp/build-company/Equipment.cs
namespace BuildCompany;

/// <summary>
/// Абстрактный базовый класс для всего оборудования
/// </summary>
public abstract class Equipment
{
    protected string name;
    protected int count;

    /// <summary>
    /// Конструктор
    /// </summary>
    /// <param name="name">Название оборудования</param>
    /// <param name="count">Количество</param>
    public Equipment(string name, int count)
    {
        this.name = name;
        this.count = count;
    }

    /// <summary>
    /// Получить название оборудования
    /// </summary>
    public string GetName()
    {
        return name;
    }

    /// <summary>
    /// Получить количество
    /// </summary>
    public int GetCount()
    {
        return count;
    }

    /// <summary>
    /// Установить название оборудования
    /// </summary>
    public void SetName(string name)
    {
        this.name = name;
    }

    /// <summary>
    /// Установить количество
    /// </summary>
    public void SetCount(int count)
    {
        this.count = count;
    }

    /// <summary>
    /// Добавить количество
    /// </summary>
    public void AddCount(int amount)
    {
        count += amount;
        Console.WriteLine($"Added {amount} units. New count: {count}");
    }

    /// <summary>
    /// Удалить количество
    /// </summary>
    public void RemoveCount(int amount)
    {
        if (count >= amount)
        {
            count -= amount;
            Console.WriteLine($"Removed {amount} units. New count: {count}");
        }
        else
        {
            Console.WriteLine($"Cannot remove {amount} units. Only {count} available.");
        }
    }

    /// <summary>
    /// Абстрактный метод для отображения информации об оборудовании
    /// </summary>
    public abstract void DisplayInfo();
}
